agent/common.py

```diff
@@ -0,0 +1,243 @@
+    return host
+
+
+def _is_ipv4(potential_ip: str) -> bool:
+    """check if the potential_ip is a valid ipv4.
+
+    Args:
+        potential_ip: string.
+
+    Returns:
+        - boolean.
+    """
+    ip, _ = _split_ipv4(potential_ip)
+    try:
+        ipaddress.IPv4Address(ip)
+        return True
+    except ValueError:
+        return False
+
+
+def _split_ipv4(potential_ip: str) -> tuple[str, str | None]:
+    """split the potential_ip to get the ip and the port if existed.
+
+    Args:
+        potential_ip: string.
+
+    Returns:
+        - ip, port.
+    """
+    ip = potential_ip
+    port = None
+    if ":" in potential_ip:
+        ip, port = potential_ip.split(":", maxsplit=1)
+    return ip, port
+
+
+def _is_ipv6(potential_ip: str) -> bool:
+    """check if the potential_ip is a valid ipv6.
+
+    Args:
+        potential_ip: string.
+
+    Returns:
+        - boolean.
+    """
+    try:
+        ipaddress.IPv6Address(potential_ip)
+        return True
+    except ValueError:
+        return False
+
+
+def build_vuln_location(
+    target_url: str,
+    file_path: str | None = None,
+    file_content: bytes | None = None,
+    selector: str | None = None,
+    content_url: str | None = None,
+) -> agent_report_vulnerability_mixin.VulnerabilityLocation:
+    """Build VulnerabilityLocation based on the asset.
+
+    Args:
+        target_url: The URL
+        file_path: The path of the file
+        file_content: The content of the file
+        selector: The message selector
+        content_url: The content url of the file
+    Returns:
+        The vulnerability location object.
+    """
+
+    asset: (
+        ipv4_asset.IPv4
+        | ipv6_asset.IPv6
+        | domain_asset.DomainName
+        | file_asset.File
+        | android_aab_asset.AndroidAab
+        | android_apk_asset.AndroidApk
+        | ios_ipa_asset.IOSIpa
+    )
+
+    if file_path is not None:
+        if selector == "v3.asset.file.android.apk":
+            asset = android_apk_asset.AndroidApk(
+                path=file_path, content=file_content, content_url=content_url
+            )
+        elif selector == "v3.asset.file.android.aab":
+            asset = android_aab_asset.AndroidAab(
+                path=file_path, content=file_content, content_url=content_url
+            )
+        elif selector == "v3.asset.file.ios.ipa":
+            asset = ios_ipa_asset.IOSIpa(
+                path=file_path, content=file_content, content_url=content_url
+            )
+        else:
+            asset = file_asset.File(path=file_path, content=file_content)
+        return agent_report_vulnerability_mixin.VulnerabilityLocation(
+            asset=asset,
+            metadata=[
+                agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                    metadata_type=agent_report_vulnerability_mixin.MetadataType.FILE_PATH,
+                    value=file_path,
+                )
+            ],
+        )
```

Comment:
if ... else ... to make the logic clearer please.

Reply:
updated